### PR TITLE
[SPARK-20909][SQL] Add build-int SQL function - DAYOFWEEK

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

Add build-int SQL function - DAYOFWEEK

## How was this patch tested?

unit tests

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/inputs/datetime.sql [hunk 1]
```diff
@@ -6,3 +6,5 @@ select current_date = current_date(), current_timestamp = current_timestamp();
 select to_date(null), to_date('2016-12-31'), to_date('2016-12-31', 'yyyy-MM-dd');
 
 select to_timestamp(null), to_timestamp('2016-12-31 00:12:00'), to_timestamp('2016-12-31', 'yyyy-MM-dd');
+
+select dayofweek('2007-02-03'), dayofweek('2009-07-30'), dayofweek('2017-05-27'), dayofweek(null), dayofweek('1582-10-15 13:10:15');
```
Comment (ueshin) [inline sql/core/src/test/resources/sql-tests/inputs/datetime.sql]: nit: Can you add a line break at the end of file?

**Review by ueshin — COMMENTED** [review]
LGTM except for a minor comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 2]
```diff
@@ -402,6 +402,44 @@ case class DayOfMonth(child: Expression) extends UnaryExpression with ImplicitCa
...
+
+  override def dataType: DataType = IntegerType
+
+  @transient private lazy val c = {
+    Calendar.getInstance(DateTimeUtils.getTimeZone("UTC"))
+  }
+
+  override protected def nullSafeEval(date: Any): Any = {
+    c.setTimeInMillis(date.asInstanceOf[Int] * 1000L * 3600L * 24L)
+    c.get(Calendar.DAY_OF_WEEK)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: In `WeekOfYear`, we set `Calendar.MONDAY` as the first day of a week. Here seems we assume it's `Calendar.SUNDAY`. Is there any conflict we will encounter?

Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Keep pace with [Hive's DayOfWeek
](https://github.com/apache/hive/blob/59539885725a96cca4b3f0759a5b26e0d8198dc8/ql/src/java/org/apache/hadoop/hive/ql/udf/UDFDayOfWeek.java#L55).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 3]
```diff
@@ -404,6 +404,44 @@ case class DayOfMonth(child: Expression) extends UnaryExpression with ImplicitCa
 
 // scalastyle:off line.size.limit
 @ExpressionDescription(
+  usage = "_FUNC_(date) - Returns the weekday index for date/timestamp (1 = Sunday, 2 = Monday, ..., 7 = Saturday).",
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: As Sunday, Saturday are included, it is not only weekday. `Returns the day of the week ...`.

**Comment by viirya** [comment]
LGTM except for one comment about the function description.

**Comment by ueshin** [comment]
Thanks! Merging to master.

**Comment by sergiobilello-eb** [comment]
This function is not reported in the API documentation: https://spark.apache.org/docs/2.2.0/api/java/index.html?org/apache/spark/sql/functions.html or https://spark.apache.org/docs/2.1.0/api/java/index.html?org/apache/spark/sql/functions.html ?
How is it possible? I am trying to use it from spark-sql on a spark 2.1.0 cluster 
Thanks

**Comment by gatorsmile** [comment]
@sergiobilello-eb This is just a SQL function. You can call it in SQL interface or using the `df.select(expr("dayofweek('2009-07-30')"))`. It is not part of the DataFrame functions. You can submit a PR or report it as an issue to add such an API.

**Comment by sergiobilello-eb** [comment]
thanks @gatorsmile :)

**Comment by sergiobilello-eb** [comment]
```spark-sql> select dayofweek('2007-02-03'), dayofweek('2009-07-30'), dayofweek('2017-05-27'), dayofweek(null), dayofweek('1582-10-15 13:10:15');
17/11/10 16:08:23 INFO SparkSqlParser: Parsing command: select dayofweek('2007-02-03'), dayofweek('2009-07-30'), dayofweek('2017-05-27'), dayofweek(null), dayofweek('1582-10-15 13:10:15')
17/11/10 16:08:23 INFO HiveMetaStore: 0: get_database: default
17/11/10 16:08:23 INFO audit: ugi=sergio.bilello	ip=unknown-ip-addr	cmd=get_database: default
17/11/10 16:08:23 INFO HiveMetaStore: 0: get_database: default
17/11/10 16:08:23 INFO audit: ugi=sergio.bilello	ip=unknown-ip-addr	cmd=get_database: default
17/11/10 16:08:23 INFO HiveMetaStore: 0: get_function: default.dayofweek
17/11/10 16:08:23 INFO audit: ugi=sergio.bilello	ip=unknown-ip-addr	cmd=get_function: default.dayofweek
Error in query: Undefined function: 'dayofweek'. This function is neither a registered temporary function nor a permanent function registered in the database 'default'.; line 1 pos 7```

**Comment by gatorsmile** [comment]
https://issues.apache.org/jira/browse/SPARK-20909 

This is not part of 2.2. Based on JIRA, it will be included in 2.3

**Comment by sergiobilello-eb** [comment]
thanks @gatorsmile! Do you suggest any workaround until then? I mean not rebuilding spark with that patch.... Can I register my UDF that contains that logic?

**Comment by gatorsmile** [comment]
Yeah, you always can implement such a UDF.

**Pull request closed without merging**